src/Callable.h

```diff
@@ -0,0 +1,357 @@
+        return Dims;
+    }
+};
+
+}  // namespace Internal
+
+class Callable {
+private:
+    friend class Pipeline;
+    friend struct CallableContents;
+    friend class PythonBindings::PyCallable;
+
+    Internal::IntrusivePtr<CallableContents> contents;
+
+    // ---------------------------------
+
+    // This value is constructed so we can do the necessary runtime check
+    // with a single 16-bit compare. It's designed to to the minimal checking
+    // necessary to ensure that the arguments are well-formed, but not necessarily
+    // "correct"; in particular, it deliberately skips checking type-and-dim
+    // of Buffer arguments, since the generated code has assertions to check
+    // for that anyway.
+    using QuickCallCheckInfo = uint16_t;
+
+    static constexpr QuickCallCheckInfo _make_qcci(uint8_t code, uint8_t bits) {
+        return (((uint16_t)code) << 8) | (uint16_t)bits;
+    }
+
+    static constexpr QuickCallCheckInfo make_scalar_qcci(halide_type_t t) {
+        return _make_qcci(t.code, t.bits);
+    }
+
+    static constexpr QuickCallCheckInfo make_buffer_qcci() {
+        constexpr uint8_t fake_bits_buffer_cci = 3;
+        return _make_qcci(halide_type_handle, fake_bits_buffer_cci);
+    }
+
+    static constexpr QuickCallCheckInfo make_ucon_qcci() {
+        constexpr uint8_t fake_bits_ucon_cci = 5;
+        return _make_qcci(halide_type_handle, fake_bits_ucon_cci);
+    }
+
+    template<typename T>
+    static constexpr QuickCallCheckInfo make_qcci() {
+        using T0 = typename std::remove_const<typename std::remove_reference<T>::type>::type;
+        if constexpr (std::is_same<T0, JITUserContext *>::value) {
+            return make_ucon_qcci();
+        } else if constexpr (Internal::IsHalideBuffer<T0>::value) {
+            // Don't bother checking type-and-dimensions here (the callee will do that)
+            return make_buffer_qcci();
+        } else if constexpr (std::is_arithmetic<T0>::value || std::is_pointer<T0>::value) {
+            return make_scalar_qcci(halide_type_of<T0>());
+        } else {
+            // static_assert(false) will fail all the time, even inside constexpr,
+            // but gating on sizeof(T) is a nice trick that ensures we will always
+            // fail here (since no T is ever size 0).
+            static_assert(!sizeof(T), "Illegal type passed to Callable.");
+        }
+    }
+
+    template<typename... Args>
+    static constexpr std::array<QuickCallCheckInfo, sizeof...(Args)> make_qcci_array() {
+        return std::array<QuickCallCheckInfo, sizeof...(Args)>{make_qcci<Args>()...};
+    }
+
+    // ---------------------------------
+
+    // This value is constructed so we can do a complete type-and-dim check
+    // of Buffers, and is used for the make_std_function() method, to ensure
+    // that if we specify static type-and-dims for Buffers, the ones we specify
+    // actually match the underlying code. We take horrible liberties with halide_type_t
```

Comment:
please describe the liberties in the comment. I see some repurposing of the lanes field below?

Reply:
Done.